presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPullConstantsAboveGroupBy.java

```diff
@@ -0,0 +1,159 @@
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+
+public class TestPullConstantsAboveGroupBy
+        extends BaseRuleTest
+{
+    @Test
+    public void testNoConstGroupingKeysDoesNotFire()
+    {
+        tester().assertThat(new PullConstantsAboveGroupBy())
+                .on(p -> p.aggregation(ab -> ab
+                        .source(
+                                p.project(
+                                        Assignments.builder()
+                                                .put(p.variable("COL"), p.rowExpression("COL"))
+                                                .put(p.variable("CONST_COL"), p.rowExpression("1"))
+                                                .build(),
+                                        p.values(p.variable("COL"))))
+                        .addAggregation(p.variable("AVG", DOUBLE), p.rowExpression("avg(COL)"))
+                        .singleGroupingSet(p.variable("COL"))))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testMultipleGroupingSetsDoesNotFire()
+    {
+        tester().assertThat(new PullConstantsAboveGroupBy())
+                .on(p -> p.aggregation(ab -> ab
+                        .source(
+                                p.project(
+                                        Assignments.builder()
+                                                .put(p.variable("COL"), p.rowExpression("COL"))
+                                                .put(p.variable("CONST_COL"), p.rowExpression("1"))
+                                                .build(),
+                                        p.values(p.variable("COL"))))
+                        .addAggregation(p.variable("AVG", DOUBLE), p.rowExpression("avg(COL)"))
+                        .groupingSets(
+                                groupingSets(ImmutableList.of(p.variable("COL")), 2, ImmutableSet.of(0)))))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testRuleDisabledDoesNotFire()
+    {
+        RuleTester tester = new RuleTester(ImmutableList.of(), ImmutableMap.of("optimize_constant_grouping_keys", "false"));
+
+        tester.assertThat(new PullConstantsAboveGroupBy())
+            .on(p -> p.aggregation(ab -> ab
+                    .source(
+                            p.project(
+                                    Assignments.builder()
+                                            .put(p.variable("COL"), p.rowExpression("COL"))
+                                            .put(p.variable("CONST_COL"), p.rowExpression("1"))
+                                            .build(),
+                                    p.values(p.variable("COL"))))
+                    .addAggregation(p.variable("AVG", DOUBLE), p.rowExpression("avg(COL)"))
+                    .singleGroupingSet(p.variable("CONST_COL"), p.variable("COL"))))
+            .doesNotFire();
+    }
+
+    @Test
+    public void test()
```

Comment:
give this a descriptive name e.g. testSingleConstColumn

Reply:
done